gateway/tests/gateway_test.rs

```diff
@@ -1143,3 +1147,312 @@ fn test_apply_msg_match_target_subnet() {
+    let r = h.submit_cron(&mut rt, submitter, checkpoint.clone());
+    assert!(r.is_ok());
+    let submission = get_epoch_submissions(&mut rt, epoch).unwrap();
+    assert_eq!(
+        submission
+            .get_submission(rt.store(), &checkpoint.hash().unwrap())
+            .unwrap()
+            .unwrap(),
+        checkpoint
+    );
+    let st: State = rt.get_state();
+    assert_eq!(st.last_cron_executed_epoch, *DEFAULT_GENESIS_EPOCH); // not executed yet
+
+    // fourth submission, executed
+    let submitter = Address::new_id(3);
+    let r = h.submit_cron(&mut rt, submitter, checkpoint.clone());
+    assert!(r.is_ok());
+    let submission = get_epoch_submissions(&mut rt, epoch);
+    assert!(submission.is_none());
+    let st: State = rt.get_state();
+    assert_eq!(st.last_cron_executed_epoch, epoch); //
+}
+
+fn storable_msg(nonce: u64) -> StorableMsg {
+    StorableMsg {
+        from: IPCAddress::new(&ROOTNET_ID, &Address::new_id(10)).unwrap(),
+        to: IPCAddress::new(&ROOTNET_ID, &Address::new_id(20)).unwrap(),
+        method: 0,
+        params: Default::default(),
+        value: Default::default(),
+        nonce,
+    }
+}
+
+#[test]
+fn test_submit_cron_abort() {
+    let (h, mut rt) = setup_root();
+
+    setup_membership(&h, &mut rt);
+
+    let epoch = *DEFAULT_GENESIS_EPOCH + *DEFAULT_CRON_PERIOD;
+
+    // first submission
+    let submitter = Address::new_id(0);
+    let checkpoint = CronCheckpoint {
+        epoch,
+        top_down_msgs: vec![],
+    };
+    let r = h.submit_cron(&mut rt, submitter, checkpoint.clone());
+    assert!(r.is_ok());
+
+    // second submission
+    let submitter = Address::new_id(1);
+    let checkpoint = CronCheckpoint {
+        epoch,
+        top_down_msgs: vec![storable_msg(1)],
+    };
+    let r = h.submit_cron(&mut rt, submitter, checkpoint.clone());
+    assert!(r.is_ok());
+
+    // third submission
+    let submitter = Address::new_id(2);
+    let checkpoint = CronCheckpoint {
+        epoch,
+        top_down_msgs: vec![storable_msg(1), storable_msg(2)],
+    };
+    let r = h.submit_cron(&mut rt, submitter, checkpoint.clone());
+    assert!(r.is_ok());
+
+    // fourth submission, aborted
+    let submitter = Address::new_id(3);
+    let checkpoint = CronCheckpoint {
+        epoch,
+        top_down_msgs: vec![storable_msg(1), storable_msg(2), storable_msg(3)],
+    };
+    let r = h.submit_cron(&mut rt, submitter, checkpoint.clone());
+    assert!(r.is_ok());
+
+    // check aborted
+    let st: State = rt.get_state();
+    assert_eq!(st.last_cron_executed_epoch, *DEFAULT_GENESIS_EPOCH); // not executed yet
+    let submission = get_epoch_submissions(&mut rt, epoch).unwrap();
+    for i in 0..4 {
+        assert_eq!(
+            submission
+                .has_submitted(rt.store(), &Address::new_id(i))
+                .unwrap(),
+            false
+        );
+    }
+}
+
+#[test]
+fn test_submit_cron_sequential_execution() {
```

Comment:
Add a few tests to see that the message is actually sent.

Reply:
added